Add ssl argument to Mongo connection

babs opened the pull request:
### What does this PR do?
Add ssl argument to Mongo connection

waynew commented on salt/states/mongodb_user.py:
```diff
@@ -58,6 +59,9 @@ def present(
     authdb
         The database in which to authenticate
 
+    ssl
+        Use ssl/tls to connect
+
```
**suggestion** add versionadded, e.g.

```suggestion
    ssl
        Use ssl/tls to connect

        .. versionadded:: Phosphorus

```

waynew reviewed (changes requested):
I really like this addition, for obvious reasons :smile: 

We *do* require test coverage for code that's changed, and we don't really have that here. Is that something that you'd be wiling to add?

If not, no big deal - we'll just add the help-wanted label and will be able to merge when we can write tests for it.

Also this will need a [changelog entry](https://docs.saltstack.com/en/latest/topics/development/changelog.html)

babs commented:
I'll call for help on that :/

waynew committed:
Add tests + ssl arg for mongodb_user state

These tests are not *super* meaningful but they do at least ensure that
ssl the arg correctly gets passed into the MongoClient. Still left to do
is add tests + ssl arg for the mongodb_database state, the mongodb
module, as well as updating all of the documentation for these modules.

We also want to add a few functional tests that ensure that the SSL
argument is correctly used with an actual mongodb server.

waynew committed:
Add docs & ssl arg to mongodb_database state

... as well as the necessary modules, natch.

waynew committed:
Add docs for mongodb module

waynew committed:
Add tests for passing ssl argument

Would still like to add functional tests against an actual mongodb.

waynew committed:
pass ssl to mongo top as well

waynew committed:
Add ssl argument to mongo pillar

waynew committed:
actually test all the cases for mongo top

waynew committed:
Add ssl argument to MongoClient for mongo returner

waynew committed:
Add ssl argument to MongoClient for mongo future returner

waynew committed:
Also add verify_ssl argument

Without verify_ssl it's a bit complicated to test, and if someone wants
to use self-signed certs in their environment without actually
installing the certs or making a CA, that should also be up to them.

waynew committed:
add changelog

waynew committed:
Update docs for module and states

waynew commented:
I went ahead and implemented the changes+tests. Just re-merged the latest changes from master in, so :crossed_fingers: hopefully all the tests pass now.

babs commented:
@waynew done :)
sorry for the delay.

babs committed:
Fix according to pre-commit

dwoz commented:
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

babs commented:
It's still relevant unless another PR has been merged that adds ssl to mongodb...

dwoz commented:
@babs Please rebase and fix the conflicts.

babs committed:
Pre-commit black + isort + pyupgrade

babs commented:
@waynew might need a second pair of eyes and possibly hands to check the rebase :)

babs committed:
Adapt tests

babs committed:
Adapt auth, authenticate on db fails

babs commented:
@waynew might need your help on mongo ssl tests

babs commented:
@dwoz do you know somebody that could give me a hand on fixing the instace tests ?

babs commented:
@dwoz do you know somebody that could give me a hand on fixing the instance tests ?

babs committed:
Fix tests

dmurphy18 committed:
Merge branch 'master' into mongodb-ssl-arg

babs commented:
re-run all

babs commented:
re-run all

babs committed:
Don't fail if pymongo is missing
